### PR TITLE
Feature/firestore/ksshiraja

### PR DESCRIPTION
**Pull request by ksshiraja**
Added firestore
Moved cassandra to a new file

### CODE REVIEW

File: provider/cassandra.go [hunk 1]
```diff
@@ -0,0 +1,206 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
```
Comment (sdreyer) [inline provider/cassandra.go:1]: If you split files again, should make different PRs for the new code and file refactor. Same goes for @RiddhiBagadiaa , but this is fine for now

File: provider/online_test.go [hunk 2]
```diff
@@ -39,24 +39,34 @@ func TestOnlineStores(t *testing.T) {
...
 
+	//Cassandra
 	cassandraAddr := "localhost:9042"
 	cassandraConfig := &CassandraConfig{
 		Addr:        cassandraAddr,
 		Consistency: gocql.One,
 	}
 
+	//Firestore
+	projectID := "snappy-density-352219"
```
Comment (sdreyer) [inline provider/online_test.go:65]: Need to import the API Key too right? Well talk later about how to add the api key to testing

**Comment by sdreyer** [comment]
You should rebase this branch. Some changes were made to fix those failing tests. Should be resolved after rebase

File: client/src/featureform/register.py [hunk 3]
```diff
@@ -289,6 +289,40 @@ def register_redis(self,
         self.__resources.append(provider)
         return OnlineProvider(self, provider)
 
+    def register_firestore(self,
+                       name: str,
+                       description: str = "",
+                       team: str = "",
+                       collection: str = "",
+                       project_id: str = "",
+                       credentials: bytearray = []
```
Comment (sdreyer) [inline client/src/featureform/register.py:298]: Change this to accept a file location, then json.dumps the contents of the file as the parameter for the FirestoreConfig

File: client/src/featureform/register.py [hunk 4]
```diff
@@ -289,6 +289,40 @@ def register_redis(self,
...
+
+    def register_cassandra(self,
+                       name: str,
+                       description: str = "",
+                       team: str = "",
+                       host: str = "0.0.0.0",
+                       port: int = 9042,
+                       keyspace: str = "",
+                       consistency: str = "ONE"):
+        config = CassandraConfig(host=host, port=port, keyspace=keyspace, consistency=consistency)
```
Comment (sdreyer) [inline client/src/featureform/register.py:317]: Does cassandra need a password?

File: provider/cassandra.go [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+	err := cassandraCluster.Consistency.UnmarshalText([]byte(options.Consistency))
+	if err != nil {
+		return nil, err
+	}
+	newSession, err := cassandraCluster.CreateSession()
+	if err != nil {
+		return nil, err
+	}
+
+	query := fmt.Sprintf("CREATE KEYSPACE IF NOT EXISTS %s WITH replication = {'class' : 'SimpleStrategy','replication_factor' : 3}", options.keyspace)
```
Comment (sdreyer) [inline provider/cassandra.go:62]: Allow replication factor to be part of the options

File: client/src/featureform/resources.py [hunk 6]
```diff
@@ -39,6 +39,48 @@ def serialize(self) -> bytes:
...
+    def software(self) -> str:
+        return "firestore"
+
+    def type(self) -> str:
+        return "FIRESTORE_ONLINE"
+
+    def serialize(self) -> bytes:
+        config = {
+            "Collection": self.collection,
+            "Project ID": self.project_id,
```
Comment (sdreyer) [inline client/src/featureform/resources.py:58]: This Project ID spelling needs to match the ProjectID in the go struct

File: provider/firestore.go [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+}
+
+func NewFirestoreOnlineStore(options *FirestoreConfig) (*firestoreOnlineStore, error) {
+	firestoreClient, err := firestore.NewClient(ctx, options.ProjectID, option.WithCredentialsJSON(options.Credentials))
+	if err != nil {
+		return nil, err
+	}
+
+	firestoreCollection := firestoreClient.Collection(options.Collection)
+	_, err = firestoreCollection.Doc("firestoreMetadata").Set(ctx, map[string]interface{}{})
```
Comment (sdreyer) [inline provider/firestore.go:57]: Change this to featureform_firestore_metadata

File: client/src/featureform/register.py [hunk 8]
```diff
@@ -289,6 +289,40 @@ def register_redis(self,
...
+        return OnlineProvider(self, provider)
+
+    def register_cassandra(self,
+                       name: str,
+                       description: str = "",
+                       team: str = "",
+                       host: str = "0.0.0.0",
+                       port: int = 9042,
+                       keyspace: str = "",
+                       consistency: str = "ONE"):
```
Comment (sdreyer) [inline client/src/featureform/register.py:316]: Change this to THREE for default

File: client/src/featureform/resources.py [hunk 9]
```diff
@@ -39,6 +39,54 @@ def serialize(self) -> bytes:
         }
         return bytes(json.dumps(config), "utf-8")
 
+@typechecked
+@dataclass
+class FirestoreConfig:
+    collection: str
+    project_id: str
+    credentials: str
```
Comment (sdreyer) [inline client/src/featureform/resources.py:47]: Change this to credentials_path to make it a bit more clear

File: provider/cassandra.go [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+	},
+	}, nil
+}
+
+func (store *cassandraOnlineStore) AsOnlineStore() (OnlineStore, error) {
+	return store, nil
+}
+
+func (store *cassandraOnlineStore) CreateTable(feature, variant string, valueType ValueType) (OnlineStoreTable, error) {
+	tableName := fmt.Sprintf("%s.table%sv%s", store.keyspace, sn.Custom(feature, "[^a-zA-Z0-9_]"), sn.Custom(variant, "[^a-zA-Z0-9_]"))
```
Comment (sdreyer) [inline provider/cassandra.go:91]: instead of v, do a double underscore if it will allow

Comment (sdreyer) [reply provider/cassandra.go:91]: Also just create a function called
GetTableName(feature, variant) string {}
that does this line so you dont have to change it in like 5 places everytime

File: provider/cassandra.go [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+	vType := cassandraTypeMap[string(valueType)]
+	key := cassandraTableKey{store.keyspace, feature, variant}
+	getTable, _ := store.GetTable(feature, variant)
+	if getTable != nil {
+		return nil, &TableAlreadyExists{feature, variant}
+	}
+
+	metadataTableName := fmt.Sprintf("%s.tableMetadata", store.keyspace)
```
Comment (sdreyer) [inline provider/cassandra.go:99]: Same here, make a 
GetMetadataTableName
and call that each time so its all in one place

File: provider/cassandra.go (hunk 10 above)
Comment (sdreyer) [reply provider/cassandra.go:91]: Also change it to "%s.featureform__%s__%s"

File: provider/provider.go [hunk 12]
```diff
@@ -58,8 +57,10 @@ func (r *RedisConfig) Deserialize(config SerializedConfig) error {
 type CassandraConfig struct {
 	keyspace    string
 	Addr        string
```
Comment (sdreyer) [inline provider/provider.go:59]: This needs to be changed to two fields host, port

File: provider/provider.go [hunk 13]
```diff
@@ -58,8 +57,10 @@ func (r *RedisConfig) Deserialize(config SerializedConfig) error {
 type CassandraConfig struct {
 	keyspace    string
```
Comment (sdreyer) [inline provider/provider.go:58]: Think this needs to be capitalized

File: client/src/featureform/resources.py [hunk 14]
```diff
@@ -39,6 +39,54 @@ def serialize(self) -> bytes:
...
+        return "firestore"
+
+    def type(self) -> str:
+        return "FIRESTORE_ONLINE"
+
+    def serialize(self) -> bytes:
+        config = {
+            "Collection": self.collection,
+            "ProjectID": self.project_id,
+            "Credentials": json.load(open(self.credentials)),
```
[the remark this answers is not in the document]
Comment (steffitan23) [reply client/src/featureform/resources.py:59]: i think the json.dumps() for the config serialized all of it?

File: client/src/featureform/register.py [hunk 15]
```diff
@@ -1,15 +1,12 @@
 # This Source Code Form is subject to the terms of the Mozilla Public
 # License, v. 2.0. If a copy of the MPL was not distributed with this
 # file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+from numpy import byte
+from .resources import ResourceState, Provider, RedisConfig, DynamodbConfig, LocalConfig, FirestoreConfig, PostgresConfig, CassandraConfig, SnowflakeConfig, RedshiftConfig, User, Location, Source, PrimaryData, SQLTable, SQLTransformation, Entity, Feature, Label, ResourceColumnMapping, TrainingSet
```
Comment (sdreyer) [inline client/src/featureform/register.py:6]: Reformat, line is very long

File: client/src/featureform/resources_test.py [hunk 16]
```diff
@@ -3,7 +3,7 @@
 # file, You can obtain one at https://mozilla.org/MPL/2.0/.
 
 import pytest
-from resources import ResourceRedefinedError, ResourceState, Provider, RedisConfig, SnowflakeConfig, PostgresConfig, RedshiftConfig, User, Provider, Entity, Feature, Label, TrainingSet, PrimaryData, SQLTable, Source, ResourceColumnMapping, DynamodbConfig, Schedule
+from resources import ResourceRedefinedError, ResourceState, Provider, RedisConfig, CassandraConfig, FirestoreConfig, SnowflakeConfig, PostgresConfig, RedshiftConfig, User, Provider, Entity, Feature, Label, TrainingSet, PrimaryData, SQLTable, Source, ResourceColumnMapping, DynamodbConfig, Schedule
```
Comment (sdreyer) [inline client/src/featureform/resources_test.py:6]: reformat with newline

File: client/src/featureform/register_test.py [hunk 17]
```diff
@@ -85,6 +85,36 @@ def test_register_user(registrar, args, expected):
...
+        "name": "firestore",
+        "description": "test",
+        "team": "featureform",
+        "collection": "abc",
+        "project_id": "abc",
+        "credentials_path": "abc"
+    },
+])
+def test_register_firestore(registrar, args):
+    registrar.register_firestore(**args)
```
Comment (ahmadnazeri) [inline client/src/featureform/register_test.py:102]: do we want to test anything here?

File: client/src/featureform/register_test.py [hunk 18]
```diff
@@ -100,15 +130,52 @@ def test_register_user(registrar, args, expected):
...
 minimal_user_args = {
     "name": "user",
     "region": "abc",
     "access_key": "abc",
     "secret_key": "abc"
 }
 
+
+@pytest.mark.parametrize("args,expected", [(minimal_user_args, User("user"))])
+def test_register_user(registrar, args, expected):
```
Comment (ahmadnazeri) [inline client/src/featureform/register_test.py:143]: what's the need for the two of the same test? Is it the same as Lines 112-117? If so, can we combine them?
